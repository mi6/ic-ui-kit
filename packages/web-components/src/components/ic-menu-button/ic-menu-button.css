@import "../../global/normalize.css";

li {
  list-style: none;
  padding: 0;
}

:host::part(button) {
  color: var(--ic-color-primary-text);
  padding: var(--ic-space-xs);
  display: flex;
  align-items: flex-start;
  gap: var(--ic-space-xs);
  border-radius: 0;
  min-height: 2.5rem;
  height: auto;
}

:host(.disabled)::part(button) {
  color: var(--ic-architectural-200) !important;
}

:host(.disabled) .menu-button-label,
:host(.disabled) .shortcut,
:host(.disabled) .menu-button-description {
  color: var(--ic-architectural-200) !important;
}

:host([variant="destructive"])::part(button) {
  color: var(--ic-action-destructive);
}

:host::part(button):hover {
  background-color: var(--ic-action-dark-bg-hover);
}

:host::part(button):focus-visible,
:host::part(button):focus {
  color: var(--ic-color-white-text);
  background-color: var(--ic-focus-blue);
  box-shadow: none;
}

:host([variant="destructive"])::part(button):focus-visible,
:host([variant="destructive"])::part(button):focus {
  background-color: var(--ic-action-destructive);
}

ic-button:focus-within .menu-button-label,
ic-button:focus-within .shortcut,
ic-button:focus-within .menu-button-description {
  color: var(--ic-color-white-text);
}

:host::part(button):active {
  background-color: var(--ic-action-dark-bg-active);
  color: var(--ic-color-primary-text) !important;
}

.menu-button-info {
  display: flex;
  flex-direction: column;
  align-items: baseline;
  white-space: pre-line;
  text-align: left;
}

.menu-labels {
  display: flex;
  align-items: baseline;
  gap: 8px;
}

.menu-button-description {
  color: var(--ic-color-secondary-text);
}

.shortcut {
  color: var(--ic-color-tertiary-text);
  display: flex;
  flex: min-content;
}

:host([variant="destructive"]) .menu-button-label {
  text-transform: uppercase;
  font-weight: 600;
  letter-spacing: 0.025em;
}

.check-icon,
.icon,
.submenu-icon {
  display: flex;
  flex: var(--ic-space-lg);
}

.check-icon svg,
.submenu-icon svg {
  height: var(--ic-space-lg);
  width: var(--ic-space-lg);
}

.check-icon.hide {
  visibility: hidden;
}

.visually-hidden {
  position: absolute;
  left: -625rem;
  top: auto;
  width: 1px;
  height: 1px;
  overflow: hidden;
}
