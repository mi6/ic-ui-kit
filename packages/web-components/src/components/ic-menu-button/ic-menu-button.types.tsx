export type IcMenuButtonVariants = "toggle" | "destructive" | "default";
