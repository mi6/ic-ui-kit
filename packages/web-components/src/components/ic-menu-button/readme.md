# ic-menu-button



<!-- Auto Generated Below -->


## Properties

| Property             | Attribute             | Description                                                                                                                                    | Type                                                                                                                                                                                     | Default     |
| -------------------- | --------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `description`        | `description`         | The description displayed in the menu button, below the label.                                                                                 | `string`                                                                                                                                                                                 | `undefined` |
| `disabled`           | `disabled`            | If `true`, the menu button will be in disabled state.                                                                                          | `boolean`                                                                                                                                                                                | `false`     |
| `href`               | `href`                | The URL that the link points to. This will render the menu button as an "a" tag.                                                               | `string`                                                                                                                                                                                 | `undefined` |
| `hreflang`           | `hreflang`            | The human language of the linked URL.                                                                                                          | `string`                                                                                                                                                                                 | `undefined` |
| `keyboardShortcut`   | `keyboard-shortcut`   | The label describing the keyboard shortcut for a menu button's action.                                                                         | `string`                                                                                                                                                                                 | `undefined` |
| `label` _(required)_ | `label`               | The label to display in the menu button.                                                                                                       | `string`                                                                                                                                                                                 | `undefined` |
| `referrerpolicy`     | `referrerpolicy`      | How much of the referrer to send when following the link.                                                                                      | `"" \| "no-referrer" \| "no-referrer-when-downgrade" \| "origin" \| "origin-when-cross-origin" \| "same-origin" \| "strict-origin" \| "strict-origin-when-cross-origin" \| "unsafe-url"` | `undefined` |
| `rel`                | `rel`                 | The relationship of the linked URL as space-separated link types.                                                                              | `string`                                                                                                                                                                                 | `undefined` |
| `submenuTriggerFor`  | `submenu-trigger-for` | This references the popover menu instance that the menu button is a trigger for. If this prop is set, then the variant will always be default. | `string`                                                                                                                                                                                 | `undefined` |
| `target`             | `target`              | The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe).                                            | `string`                                                                                                                                                                                 | `undefined` |
| `variant`            | `variant`             | The variant of the menu button.                                                                                                                | `"default" \| "destructive" \| "toggle"`                                                                                                                                                 | `"default"` |


## Events

| Event                   | Description | Type                                                   |
| ----------------------- | ----------- | ------------------------------------------------------ |
| `handleMenuButtonClick` |             | `CustomEvent<{ label: string; hasSubMenu: boolean; }>` |


## Slots

| Slot     | Description                                                  |
| -------- | ------------------------------------------------------------ |
| `"icon"` | Content will be placed to the left of the menu button label. |


## Dependencies

### Depends on

- [ic-typography](../ic-typography)
- [ic-button](../ic-button)

### Graph
```mermaid
graph TD;
  ic-menu-button --> ic-typography
  ic-menu-button --> ic-button
  ic-button --> ic-loading-indicator
  ic-button --> ic-tooltip
  ic-loading-indicator --> ic-typography
  ic-tooltip --> ic-typography
  style ic-menu-button fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


