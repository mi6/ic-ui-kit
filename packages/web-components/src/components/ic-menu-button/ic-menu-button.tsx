import {
  Component,
  Element,
  Host,
  Prop,
  h,
  State,
  Event,
  EventEmitter,
} from "@stencil/core";
import {
  isSlotUsed,
  onComponentRequiredPropUndefined,
} from "../../utils/helpers";
import { IcMenuButtonVariants } from "./ic-menu-button.types";
import Check from "../../assets/check-icon.svg";
import Chevron from "../../assets/chevron-icon.svg";

/**
 * @slot icon - Content will be placed to the left of the menu button label.
 */

@Component({
  tag: "ic-menu-button",
  styleUrl: "ic-menu-button.css",
  shadow: {
    delegatesFocus: true,
  },
})
export class MenuButton {
  @Element() el: HTMLIcMenuButtonElement;

  /**
   * The variant of the menu button.
   */
  @Prop({ mutable: true, reflect: true }) variant: IcMenuButtonVariants =
    "default";
  /**
   * The label to display in the menu button.
   */
  @Prop() label!: string;
  /**
   * The description displayed in the menu button, below the label.
   */
  @Prop() description?: string;
  /**
   * If `true`, the menu button will be in disabled state.
   */
  @Prop() disabled?: boolean = false;
  /**
   * The URL that the link points to. This will render the menu button as an "a" tag.
   */
  @Prop() href?: string;
  /**
   * The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe).
   */
  @Prop() target?: string;
  /**
   * The relationship of the linked URL as space-separated link types.
   */
  @Prop() rel?: string;
  /**
   * The human language of the linked URL.
   */
  @Prop() hreflang?: string;
  /**
   * How much of the referrer to send when following the link.
   */
  @Prop() referrerpolicy?: ReferrerPolicy;
  /**
   * The label describing the keyboard shortcut for a menu button's action.
   */
  @Prop() keyboardShortcut?: string;
  /**
   * This references the popover menu instance that the menu button is a trigger for. If this prop is set, then the variant will always be default.
   */
  @Prop() submenuTriggerFor?: string;

  /**
   * @internal Emitted when item loses focus.
   */
  @Event() childBlur: EventEmitter<void>;

  /**
   * @internal Emitted when the user selects a menu button that triggers a popover menu instance.
   */
  @Event() triggerPopoverMenuInstance: EventEmitter<void>;

  @State() toggleChecked: boolean = false;

  @Event() handleMenuButtonClick: EventEmitter<{
    label: string;
    hasSubMenu: boolean;
  }>;

  private handleClick = (e: Event): void => {
    e.preventDefault();
    if (this.submenuTriggerFor !== undefined) {
      this.triggerPopoverMenuInstance.emit();
    } else if (this.variant === "toggle") {
      this.toggleChecked
        ? (this.toggleChecked = false)
        : (this.toggleChecked = true);
    }
  };

  componentDidLoad(): void {
    onComponentRequiredPropUndefined(
      [{ prop: this.label, propName: "label" }],
      "Menu Button"
    );

    // This ensures that trigger menu buttons are always set to the default variant
    if (this.submenuTriggerFor !== undefined && this.variant !== "default") {
      this.variant = "default";
    }
  }

  render() {
    // A helper function that checks if a prop has been defined
    const isPropDefined = (prop: string) => {
      return prop !== undefined ? prop : null;
    };

    // A sub-component to layout the menu information correctly in ic-button
    const MenuButtonInformation = () => {
      return (
        <div class="menu-button-info">
          <div class="menu-labels">
            <ic-typography class="menu-button-label">
              {this.el.classList.contains("ic-popover-submenu-back-button") && (
                <span class="visually-hidden">Go </span>
              )}
              {this.label}
              {this.el.classList.contains("ic-popover-submenu-back-button") && (
                <span class="visually-hidden"> to previous menu</span>
              )}
            </ic-typography>
            {this.keyboardShortcut && (
              <ic-typography variant="caption" class="shortcut">
                {this.keyboardShortcut}
              </ic-typography>
            )}
          </div>
          {this.description && (
            <ic-typography class="menu-button-description" variant="caption">
              {this.description}
            </ic-typography>
          )}
        </div>
      );
    };

    //const userSetAriaLabel = this.el.shadowRoot.querySelector("li > ic-button");

    // this.el.setAttribute("aria-label",`${this.el.label}. ${userSetAriaLabel? this.el.getAttribute("aria-label") : "no aria"}`)
    // console.log("aria-label", userSetAriaLabel);
    // console.log("aria-label", this.el.getAttribute("aria-label"));

    return (
      <Host
        class={{
          ["disabled"]: this.disabled,
        }}
      >
        <li
          role={this.variant === "toggle" ? "menuitemcheckbox" : "menuitem"}
          aria-disabled={this.disabled ? true : false}
        >
          <ic-button
            disableTooltip
            fullWidth
            disabled={this.disabled}
            variant="tertiary"
            onClick={this.handleClick}
            href={isPropDefined(this.href)}
            hreflang={isPropDefined(this.hreflang)}
            target={isPropDefined(this.target)}
            rel={isPropDefined(this.rel)}
            referrerpolicy={
              this.referrerpolicy !== undefined ? this.referrerpolicy : null
            }
            aria-checked={
              this.variant === "toggle" && this.toggleChecked === true
                ? true
                : this.variant === "toggle" && this.toggleChecked === false
                ? false
                : undefined
            }
            aria-controls={
              this.submenuTriggerFor !== undefined
                ? `ic-popover-submenu-${this.submenuTriggerFor}`
                : false
            }
            aria-haspopup={
              this.submenuTriggerFor !== undefined ||
              this.el.classList.contains("ic-popover-submenu-back-button")
                ? "menu"
                : false
            }
            aria-owns={
              this.submenuTriggerFor !== undefined
                ? `ic-popover-submenu-${this.submenuTriggerFor}`
                : false
            }
          >
            {isSlotUsed(this.el, "icon") && (
              <span class="icon">
                <slot name="icon"></slot>
              </span>
            )}
            <MenuButtonInformation />
            {this.variant === "toggle" && (
              <span
                class={{ ["check-icon"]: true, ["hide"]: !this.toggleChecked }}
                aria-hidden="true"
                innerHTML={Check}
              />
            )}
            {this.submenuTriggerFor !== undefined && (
              <span
                class={{ ["submenu-icon"]: true }}
                aria-hidden="true"
                innerHTML={Chevron}
              />
            )}
          </ic-button>
        </li>
      </Host>
    );
  }
}
