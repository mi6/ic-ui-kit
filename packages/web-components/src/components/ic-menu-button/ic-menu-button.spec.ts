import { newSpecPage } from "@stencil/core/testing";
import { MenuButton } from "./ic-menu-button";

describe("menu button variants", () => {
  it("should render the default variant", async () => {
    const page = await newSpecPage({
      components: [MenuButton],
      html: `<ic-menu-button
            label="Default variant"
          />`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render a menu button with a description", async () => {
    const page = await newSpecPage({
      components: [MenuButton],
      html: `<ic-menu-button
            label="Default variant"
            description="This is the default variant of the menu button with a description"
          />`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with keyboard shortcut", async () => {
    const page = await newSpecPage({
      components: [MenuButton],
      html: `<ic-menu-button
              label="Toggle variant"
              keyboard-shortcut="Cmd+"
            />`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it("should render the toggle variant", async () => {
    const page = await newSpecPage({
      components: [MenuButton],
      html: `<ic-menu-button
            variant="toggle"
            label="Toggle variant"
          />`,
    });

    expect(page.root).toMatchSnapshot();
  });
  it("should render the destructive variant", async () => {
    const page = await newSpecPage({
      components: [MenuButton],
      html: `<ic-menu-button
            variant="destuctive"
            label="Destructive variant"
          />`,
    });

    expect(page.root).toMatchSnapshot();
  });
  it("should render a menu button that triggers a popover menu instance", async () => {
    const page = await newSpecPage({
      components: [MenuButton],
      html: `<ic-menu-button
            variant="destuctive"
            submenu-trigger-for="submenu-1"
            label="Destructive variant"
          />`,
    });

    expect(page.rootInstance.submenuTriggerFor).not.toBeUndefined();

    expect(page.rootInstance.variant).toMatch("default");
  });
  it('should emit the triggerPopoverMenuInstance event when the menu button has the prop: "submenu-trigger-for" and is clicked', async () => {
    const page = await newSpecPage({
      components: [MenuButton],
      html: `<ic-menu-button
      id="test-menu-button"
            submenu-trigger-for="submenu-1"
            label="I emit an event"
          />`,
    });

    const eventSpy = jest.fn();
    page.win.addEventListener("triggerPopoverMenuInstance", eventSpy);
    expect(page.rootInstance.submenuTriggerFor).not.toBeUndefined();
    const element = await document.getElementById("test-menu-button");

    await element.click();

    await page.waitForChanges();

    await page.rootInstance.handleClick;
  });
});
