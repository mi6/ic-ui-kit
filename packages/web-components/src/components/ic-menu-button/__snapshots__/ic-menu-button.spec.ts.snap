// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menu button variants should render a menu button with a description 1`] = `
<ic-menu-button description="This is the default variant of the menu button with a description" label="Default variant" variant="default">
  <mock:shadow-root>
    <li role="menuitem">
      <ic-button disabletooltip="" fullwidth="" variant="tertiary">
        <div class="menu-button-info">
          <div class="menu-labels">
            <ic-typography class="menu-button-label">
              Default variant
            </ic-typography>
          </div>
          <ic-typography class="menu-button-description" variant="caption">
            This is the default variant of the menu button with a description
          </ic-typography>
        </div>
      </ic-button>
    </li>
  </mock:shadow-root>
</ic-menu-button>
`;

exports[`menu button variants should render the default variant 1`] = `
<ic-menu-button label="Default variant" variant="default">
  <mock:shadow-root>
    <li role="menuitem">
      <ic-button disabletooltip="" fullwidth="" variant="tertiary">
        <div class="menu-button-info">
          <div class="menu-labels">
            <ic-typography class="menu-button-label">
              Default variant
            </ic-typography>
          </div>
        </div>
      </ic-button>
    </li>
  </mock:shadow-root>
</ic-menu-button>
`;

exports[`menu button variants should render the destructive variant 1`] = `
<ic-menu-button label="Destructive variant" variant="destuctive">
  <mock:shadow-root>
    <li role="menuitem">
      <ic-button disabletooltip="" fullwidth="" variant="tertiary">
        <div class="menu-button-info">
          <div class="menu-labels">
            <ic-typography class="menu-button-label">
              Destructive variant
            </ic-typography>
          </div>
        </div>
      </ic-button>
    </li>
  </mock:shadow-root>
</ic-menu-button>
`;

exports[`menu button variants should render the toggle variant 1`] = `
<ic-menu-button label="Toggle variant" variant="toggle">
  <mock:shadow-root>
    <li role="menuitemcheckbox">
      <ic-button disabletooltip="" fullwidth="" variant="tertiary">
        <div class="menu-button-info">
          <div class="menu-labels">
            <ic-typography class="menu-button-label">
              Toggle variant
            </ic-typography>
          </div>
        </div>
        <span aria-hidden="true" class="check-icon hide">
          svg
        </span>
      </ic-button>
    </li>
  </mock:shadow-root>
</ic-menu-button>
`;

exports[`menu button variants should render with keyboard shortcut 1`] = `
<ic-menu-button keyboard-shortcut="Cmd+" label="Toggle variant" variant="default">
  <mock:shadow-root>
    <li role="menuitem">
      <ic-button aria-label="Toggle variant, Cmd+" disabletooltip="" fullwidth="" variant="tertiary">
        <div class="menu-button-info">
          <div class="menu-labels">
            <ic-typography class="menu-button-label">
              Toggle variant
            </ic-typography>
            <ic-typography class="shortcut" variant="caption">
              Cmd+
            </ic-typography>
          </div>
        </div>
      </ic-button>
    </li>
  </mock:shadow-root>
</ic-menu-button>
`;
